subscriptions/providers/apple_in_app/__init__.py

```diff
@@ -102,61 +103,83 @@ def check_payments(self, payments: Iterable[SubscriptionPayment]):
                 raise AppleSubscriptionNotCompletedError(payment.provider_transaction_id)
 
     @classmethod
-    def _get_validated_in_app_product(cls, response: AppleVerifyReceiptResponse) -> AppleInApp:
+    def _is_receipt_valid(cls, response: AppleVerifyReceiptResponse) -> None:
         if not response.is_valid or response.receipt.bundle_id != cls.bundle_id:
             raise AppleReceiptValidationError()
-        return one(response.receipt.in_apps)
 
-    @transaction.atomic
-    def _handle_receipt(self, request: Request, payload: AppleReceiptRequest) -> Response:
-        receipt = payload.transaction_receipt
-
-        # Validate the receipt. Fetch the status and product.
-        receipt_data = self.api.fetch_receipt_data(receipt)
-        single_in_app = self._get_validated_in_app_product(receipt_data)
-
-        # Check whether this receipt is anyhow interesting:
+    def _handle_single_receipt_info(self,
+                                    user: User,
+                                    receipt_info: AppleLatestReceiptInfo) -> Optional[SubscriptionPayment]:
         with suppress(SubscriptionPayment.DoesNotExist):  # noqa (DoesNotExist seems not to be an exception for PyCharm)
             payment = SubscriptionPayment.objects.get(provider_codename=self.codename,
-                                                      provider_transaction_id=single_in_app.transaction_id)
-            # We've already handled this. And all the messages coming to us from iOS should be AFTER
-            # the client money were removed.
-            return Response(
-                SubscriptionPaymentSerializer(payment).data,
-                status=HTTP_200_OK,
-            )
+                                                      provider_transaction_id=receipt_info.transaction_id)
+
+            # User was refunded.
+            if receipt_info.cancellation_date is not None:
+                payment.subscription_end = receipt_info.cancellation_date
+                payment.save()
```

Comment:
For safety reasons, payment cannot auto-reduce subscription time, see this snippet:
```python
                    # change existing subscription duration based on payment's end
                    if self.subscription_end <= subscription.end:
                        log.warning('Payment\'s end (%s) is earlier than current subscription\'s end (%s) -> payment has no effect', self.subscription_end, subscription.end)
                    else:
                        subscription.end = self.subscription_end
```
So not only `payment.subscription_end = ...`, but maybe
```python
payment.subscription_end = payment.subscription.end = receipt_info.cancellation_date
payment.save()
payment.subscription.save()
```

Reply:
So I'll need to sort these by expiration time as well.

Probably subscription should be saved before the subscription payment, right?

I'll just raise an error for now, whenever we have cancellation dates. This will require a bit of work.